src/modelfree/common/utils.py

```diff
@@ -156,6 +161,87 @@ def make_session(graph=None):
     return sess
 
 
+class TrajectoryRecorder(VecMultiWrapper):
+    def __init__(self, venv, save_dir, use_gail_format=False, agent_indices=None):
+        VecMultiWrapper.__init__(self, venv)
+        self.save_dir = save_dir
+        self.use_gail_format = use_gail_format
+        if agent_indices is None:
+            self.agent_indices = range(self.num_agents)
+        elif isinstance(agent_indices, int):
+            self.agent_indices = [agent_indices]
+        os.makedirs(self.save_dir, exist_ok=True)
+
+        self.traj_dicts = [[defaultdict(list) for e in range(self.num_envs)]
+                           for p in self.agent_indices]
+        self.full_traj_dicts = [defaultdict(list) for p in self.agent_indices]
+        self.prev_obs = None
+        self.actions = None
+
+    def step_async(self, actions):
+        self.actions = actions
+        self.venv.step_async(actions)
+
+    def step_wait(self):
+        observations, rewards, dones, infos = self.venv.step_wait()
+        self.record_traj(self.prev_obs, self.actions, rewards, dones, infos)
+        self.prev_obs = observations
+        return observations, rewards, dones, infos
+
+    def reset(self):
+        observations = self.venv.reset()
+        self.prev_obs = observations
+        return observations
+
+    def record_traj(self, prev_obs, actions, rewards, dones, infos):
+        data_keys = ('rewards', 'actions', 'obs')
+        data_vals = (rewards, actions, prev_obs)
+        transparency_keys = ('ff', 'hid')  # we already record observations
+        iter_space = itertools.product(enumerate(self.traj_dicts), range(self.num_envs))
+        # iterate over both agents over all environments in VecEnv
+        for (dict_idx, agent_dicts), env_idx in iter_space:
+            # in dict number dict_idx, record trajectories for agent number agent_idx
+            agent_idx = self.agent_indices[dict_idx]
+            for key, val in zip(data_keys, data_vals):
+                # data_vals always have data for all agents (use agent_idx not dict_idx)
+                agent_dicts[env_idx][key].append(val[agent_idx][env_idx])
+
+            for key in transparency_keys:
+                # infos also always have data for all agents
+                if key not in infos[env_idx][agent_idx]:
+                    continue
+                agent_dicts[env_idx][key].append(infos[env_idx][agent_idx][key])
+            if dones[env_idx]:
+                if self.use_gail_format:
```

Comment:
GAIL compatibility code does not belong in TrajectoryRecorder, makes it harder to understand. Also we should really standardize on a single format for trajectory recording in our code, and do conversions as needed just before feeding into external libraries.